subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1833 @@
+		if (len == 0) {
+			return 0;
+		}
+	}
+
+	/* calculate required space if the entry contains data */
+	if (data_size) {
+		/* Leave space for delete ate */
+		if (len > ZMS_DATA_IN_ATE_SIZE) {
+			required_space = data_size + ate_size;
+		} else {
+			required_space = ate_size;
+		}
+	}
+
+	k_mutex_lock(&fs->zms_lock, K_FOREVER);
+
+	gc_count = 0;
+	while (1) {
+		if (gc_count == fs->sector_count) {
+			/* gc'ed all sectors, no extra space will be created
+			 * by extra gc.
+			 */
+			rc = -ENOSPC;
+			goto end;
+		}
+
+		/* We need to make sure that we leave the ATE at address 0x0 of the sector
+		 * empty (even for delete ATE). Otherwise, the fs->ate_wra will be decremented
+		 * after this write by ate_size and it will underflow.
+		 * So the first position of a sector (fs->ate_wra = 0x0) is forbidden for ATEs
+		 * and the second position could be written only be a delete ATE.
+		 */
+		if ((SECTOR_OFFSET(fs->ate_wra)) &&
+		    (fs->ate_wra >= (fs->data_wra + required_space)) &&
+		    (SECTOR_OFFSET(fs->ate_wra - ate_size) || !len)) {
+			rc = zms_flash_write_entry(fs, id, data, len);
+			if (rc) {
+				goto end;
+			}
+			break;
+		}
+		rc = zms_sector_close(fs);
+		if (rc) {
+			LOG_ERR("Failed to close the sector, returned = %d", rc);
+			goto end;
+		}
+		rc = zms_gc(fs);
+		if (rc) {
+			LOG_ERR("Garbage collection failed, returned = %d", rc);
+			goto end;
+		}
+		gc_count++;
+	}
+	rc = len;
+end:
+	k_mutex_unlock(&fs->zms_lock);
+	return rc;
+}
+
+int zms_delete(struct zms_fs *fs, uint32_t id)
+{
+	return zms_write(fs, id, NULL, 0);
+}
+
+ssize_t zms_read_hist(struct zms_fs *fs, uint32_t id, void *data, size_t len, uint32_t cnt)
+{
+	int rc, prev_found;
+	uint64_t wlk_addr, rd_addr, wlk_prev_addr;
+	uint32_t cnt_his;
+	struct zms_ate wlk_ate;
+	size_t ate_size;
+#ifdef CONFIG_ZMS_DATA_CRC
+	uint32_t computed_data_crc;
+#endif
+
+	if (!fs->ready) {
+		LOG_ERR("zms not initialized");
+		return -EACCES;
+	}
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	cnt_his = 0U;
+
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+	wlk_addr = fs->lookup_cache[zms_lookup_cache_pos(id)];
+
+	if (wlk_addr == ZMS_LOOKUP_CACHE_NO_ADDR) {
+		rc = -ENOENT;
+		goto err;
+	}
+#else
+	wlk_addr = fs->ate_wra;
+#endif
+
+	while (cnt_his <= cnt) {
+		wlk_prev_addr = wlk_addr;
+		/* Search for a previous valid ATE with the same ID */
+		prev_found = zms_find_ate_with_id(fs, id, wlk_addr, fs->ate_wra, &wlk_ate,
+						  &wlk_prev_addr);
+		if (prev_found < 0) {
+			return prev_found;
+		}
+		if (prev_found) {
+			cnt_his++;
+			/* wlk_prev_addr contain the ATE address of the previous found ATE. */
+			rd_addr = wlk_prev_addr;
+			/*
+			 * compute the previous ATE address in case we need to start
+			 * the research again.
+			 */
+			rc = zms_compute_prev_addr(fs, &wlk_prev_addr);
+			if (rc) {
+				return rc;
+			}
+			/* wlk_addr will be the start research address in the next loop */
+			wlk_addr = wlk_prev_addr;
+		} else {
+			wlk_addr = fs->ate_wra;
```

Comment:
This assignment seems unneded, as you now check `prev_found` instead of `wlk_addr` below.